src/evaluators/ForInStatement.js

```diff
@@ -160,30 +181,33 @@ function emitResidualLoopIfSafe(
         }
       });
       if (targetObject instanceof ObjectValue && sourceObject !== undefined) {
-        let oe = ob.values.getElements();
-        if (oe.size !== 1) ob.throwIfNotConcreteObject();
         let o;
-        for (let co of oe) o = co;
-        invariant(o !== undefined);
+        let oe = ob.values.getElements();
+        if (oe.size !== 1) {
+          o = ob;
```

Comment:
This is a new case to allow abstract objects that have not been obtained directly from the environmental model.